R/join_keys.R

```diff
@@ -0,0 +1,651 @@
+# Constructors ====
+
+#' Create a `join_keys` out of a list of `join_key_set` objects
+#'
+#' @description `r lifecycle::badge("stable")`
+#'
+#' @details - `join_keys()`: When called without arguments it will return an
+#' empty constructor.
+#' - `join_keys(x)`: When called with a single argument it will return the `join_keys`
+#' object contained in `x` (if it contains a `join_keys` object).
+#' - `join_keys(...)`: When called with a single or more `join_key_set` parameters it will
+#' create a new object.
+#'
+#' Note that join keys are created symmetrically, that is, if `dat1` and `dat2`
+#' have a join key of `col1`, then 2 join keys are created, `dat1 → dat2` and
+#' `dat2 → dat1`. The only exception is for a primary key.
+#'
+#' @param ... (optional), when no argument is given the empty constructor is called.
+#' Otherwise, when called with only one argument of type: `join_keys` or  `teal_data`
+#' it will return the `join_keys` of that object.
+#' When called with 1 or more `join_key_set` it will create a new `join_keys` with
+#' constructed from the arguments.
+#'
+#' @return `join_keys` object.
+#'
+#' @export
+#'
+#' @examples
+#' # Setting join keys ----
+#'
+#' jk <- join_keys(
+#'   join_key("dataset_A", "dataset_B", c("col_1" = "col_a")),
+#'   join_key("dataset_A", "dataset_C", c("col_2" = "col_x", "col_3" = "col_y"))
+#' )
+#' jk
+#'
+#' # or
+#' jk <- join_keys()
+#' jk[["dataset_A"]][["dataset_B"]] <- c("col_1" = "col_a")
+#' jk[["dataset_A"]][["dataset_C"]] <- c("col_2" = "col_x", "col_3" = "col_y")
+#' jk
+#'
+#' td <- teal_data(join_keys = join_keys(join_key("a", "b", "c")))
+#' join_keys(td)
+#'
+#' jk <- join_keys()
+#' join_keys(jk)
+#'
+#' jk <- join_keys()
+#' jk <- c(jk, join_keys(join_key("a", "b", "c")))
+#' jk <- c(jk, join_keys(join_key("a", "b", "c"), join_key("a", "b2", "c")))
+join_keys <- function(...) {
+  if (missing(...)) {
+    return(new_join_keys())
+  }
+  x <- rlang::list2(...)
+  if (length(x) > 1) {
+    return(join_keys.default(...))
+  }
+  UseMethod("join_keys", x[[1]])
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.join_keys <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.teal_data <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]@join_keys
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.TealData <- function(...) {
+  x <- rlang::list2(...)
+  x[[1]]$get_join_keys()
+}
+
+#' @rdname join_keys
+#' @export
+join_keys.default <- function(...) {
+  x <- rlang::list2(...)
+  # Constructor
+  res <- new_join_keys()
+  if (length(x) > 0) {
```

Comment:
The generic will call this method only if `length(...) > 1` so is this check necessary (or is it only for guarding direct calls)?

Reply:
Guarding against some edge case that probably doesn't exist.

Removed the check in 6ed79041